### PR TITLE
`Integrated code lifecycle`: Add build agent name

### PR DESCRIPTION
**Pull request by BBesrour**
# ⚠️ Breaking change: Admins need to update the `application-prod.yml` of their LocalCI build agents ‼️

#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

We want to add a name for build agents. The name should only allow lowercase letters, numbers, and hyphens.
We also want to use the name for REST requests (we previously used the member address)

(We also change the grace period for cancelling jobs to 60s. This is not related to the other changes)

### Steps for Testing

Prerequisites:
- 1 Admin

1. Navigate to Build Agents. Make sure that the page works and there are no visible errors
2. Navigate to Build agent details (by clicking on a name in the previous page). Make sure that the page works and that there are no visible errors
3. Make sure that you can pause the agent. The status of the agent should be changed to `paused`. Additionally, you can check if the agent was paused by submitting a job and it staying in the build job queue. (if using a multi node instance, make sure to pause all agents)
4. (OPTIONAL) Remove the newly added property `build-agent.short-name` from `application.yml`, start the server. The application should fail to start and the error message should be understandable

#### Exam Mode Testing

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| build-agent.model.ts | 100% | ✅ |
| build-job.model.ts | 100% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| BuildAgentInformation.java | 100% | ✅ |
| BuildJobQueueItem.java | 100% | ✅ |
| SharedQueueProcessingService.java | 78% | ✅ ❌ |
| LocalCITriggerService.java | 88% | ✅ ❌ |
| SharedQueueManagementService.java | 93% | ✅ ❌ |

### Screenshots

Old
image

New
image

## Release Notes

- **New Features**
  - Introduced `BuildAgentDTO` to encapsulate build agent details.
  - Added `BuildAgentInformation` class to represent comprehensive build agent data.
  - Enhanced user interface to display `memberAddress` alongside build agent names.

- **Improvements**
  - Updated build agent management logic for better clarity and type safety.
  - Enhanced logging for build job processing and error handling.

- **Documentation**
  - Expanded documentation for setting up multiple Artemis instances, including synchronization and configuration details.

### CODE REVIEW

File: src/main/resources/config/application.yml [hunk 1]
```diff
@@ -96,6 +96,14 @@ artemis:
...
                     default: "ghcr.io/ls1intum/artemis-cpp-docker:v1.0.0"
 
+        # The following properties are used to configure the Artemis build agent.
+        # The build agent is responsible for executing the buildJob to test student submissions.
+        build-agent:
+            # Name of the build agent. Only lowercase letters, numbers and hyphens are allowed. ([a-z0-9-]+)
+            short-name: "artemis-build-agent-1"
+
+
+
```
Comment (SimonEntholzer) [inline src/main/resources/config/application.yml:106]: What happens if I want more than one build agent? Do you need to explicitly give each a name?

Comment (BBesrour) [reply src/main/resources/config/application.yml:106]: Yes, then you would need to add this property in the `application.yml` of every build agent

File: docs/admin/setup/distributed.rst [hunk 2]
```diff
@@ -617,8 +617,12 @@ These credentials are used to clone repositories via HTTPS. You must also add th
                 container-cleanup:
                     expiry-minutes: 5                               # Time after a hanging container will automatically be removed
                     cleanup-schedule-minutes: 60                    # Schedule for container cleanup
+                build-agent:
+                    short-name: "artemis-build-agent-X"             # Short name of the build agent. This should be unique for each build agent. Only lowercase letters, numbers and hyphens are allowed.
```
Comment (krusche) [inline docs/admin/setup/distributed.rst:621]: can you add the new `display-name` here as well please?

File: src/main/java/de/tum/cit/aet/artemis/buildagent/dto/BuildJobQueueItem.java [hunk 3]
```diff
@@ -14,8 +14,8 @@
 // in the future are migrated or cleared. Changes should be communicated in release notes as potentially breaking changes.
 @JsonIgnoreProperties(ignoreUnknown = true)
 @JsonInclude(JsonInclude.Include.NON_EMPTY)
-public record BuildJobQueueItem(String id, String name, String buildAgentAddress, long participationId, long courseId, long exerciseId, int retryCount, int priority,
-        BuildStatus status, RepositoryInfo repositoryInfo, JobTimingInfo jobTimingInfo, BuildConfig buildConfig, ResultDTO submissionResult) implements Serializable {
+public record BuildJobQueueItem(String id, String name, String buildAgentName, String buildAgentAddress, long participationId, long courseId, long exerciseId, int retryCount,
```
Comment (krusche) [inline src/main/java/de/tum/cit/aet/artemis/buildagent/dto/BuildJobQueueItem.java:17]: I suggest to create a `BuildAgent` record and store `shortName`, `displayName` and `address` in it. This allows better reuse and avoid too long parameter lists

**Review by krusche — CHANGES_REQUESTED** [review]
see my inline feedback

File: docs/admin/setup/distributed.rst (hunk 2 above)
Comment (florian-glombik) [inline docs/admin/setup/distributed.rst:621]: Do I get it right that the default config fails due to the "X" to ensure the value has to be changed?

File: src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java [hunk 4]
```diff
@@ -114,9 +114,12 @@ public class SharedQueueProcessingService {
      */
     private final AtomicBoolean processResults = new AtomicBoolean(true);
 
-    @Value("${artemis.continuous-integration.pause-grace-period-seconds:15}")
+    @Value("${artemis.continuous-integration.pause-grace-period-seconds:60}")
```
Comment (florian-glombik) [inline src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java:117]: Is there a particular reason for the definition of the grace-period? Why is it increased?

Comment (florian-glombik) [inline src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java:117]: Is there a particular reason for the definition of the grace-period? Why is it increased?

File: src/test/resources/config/application.yml [hunk 5]
```diff
@@ -74,6 +74,8 @@ artemis:
                     default: "~~invalid~~"
                 c_plus_plus:
                     default: "~~invalid~~"
+        build-agent:
+            short-name: "artemis-build-agent-test"
```
Comment (florian-glombik) [inline src/test/resources/config/application.yml:78]: Is this aent intentonally named "test"?

File: src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java (hunk 4 above)
Comment (BBesrour) [reply src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java:117]: Yes, we wanted to increase the default grace period to 60 seconds. This is really related to this PR but I didnt want to open another PR just for this :)

File: docs/admin/setup/distributed.rst (hunk 2 above)
Comment (BBesrour) [reply docs/admin/setup/distributed.rst:621]: Yes. Do you think it's better if I put a valid value there?

File: src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java [hunk 6]
```diff
@@ -26,6 +26,7 @@
 import jakarta.annotation.PostConstruct;
 import jakarta.annotation.PreDestroy;
 
+import org.apache.commons.lang3.StringUtils;
```
[the remark this answers is not in the document]
Comment (BBesrour) [reply src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java:29]: What if buildAgentDisplayName is null?

**Review by SimonEntholzer — CHANGES_REQUESTED** [review]
image
I think the build agents should have different names, maybe something isn't set up correctly yet

**Comment by BBesrour** [comment]
@SimonEntholzer It can be that the new config isnt deployed yet. Will check with the admins

File: src/main/java/de/tum/cit/aet/artemis/core/web/admin/AdminBuildJobQueueResource.java [hunk 7]
```diff
@@ -95,8 +95,8 @@ public ResponseEntity<List<BuildAgentInformation>> getBuildAgentSummary() {
     @GetMapping("build-agent")
     public ResponseEntity<BuildAgentInformation> getBuildAgentDetails(@RequestParam String agentName) {
         log.debug("REST request to get information on build agent {}", agentName);
-        BuildAgentInformation buildAgentDetails = localCIBuildJobQueueService.getBuildAgentInformation().stream().filter(agent -> agent.name().equals(agentName)).findFirst()
-                .orElse(null);
+        BuildAgentInformation buildAgentDetails = localCIBuildJobQueueService.getBuildAgentInformation().stream().filter(agent -> agent.buildAgent().name().equals(agentName))
+                .findFirst().orElse(null);
         return ResponseEntity.ok(buildAgentDetails);
```
Comment (iyannsch) [inline src/main/java/de/tum/cit/aet/artemis/core/web/admin/AdminBuildJobQueueResource.java:100]: I know you didn't add this code, but maybe we can still discuss its validity :) As I see it here, we'd return `ResponseEntity.ok(null)` when the respective agent is not found, right? This seems like wrapping an error value in a success code

File: src/main/webapp/app/localci/build-agents/build-agent-summary/build-agent-summary.component.html [hunk 8]
```diff
@@ -75,7 +83,7 @@ <h3 id="build-agents-heading" jhiTranslate="artemisApp.buildAgents.summary"></h3
                         <ng-template ngx-datatable-cell-template let-value="value" let-row="row">
                             <span style="margin-right: 20px">{{ value }}</span>
                             @if (value > 0) {
-                                <button class="btn btn-danger btn-sm" (click)="cancelAllBuildJobs(row.name)">
+                                <button class="btn btn-danger btn-sm" (click)="cancelAllBuildJobs(row.buildAgent!.name)">
```
Comment (iyannsch) [inline src/main/webapp/app/localci/build-agents/build-agent-summary/build-agent-summary.component.html:86]: How about using `row.buildAgend` (might be `undefined`) as an argument here? We could then do the null-check in the `cancelAllBuildJobs` function and gather the name if it's existent?

**Review by iyannsch — CHANGES_REQUESTED** [review]
Nice changes, looking good! I added a small suggestion which reduces the usage of `non-undefined` assumptions 👍

**Review by iyannsch — APPROVED** [review]
Thanks for incorporating my change requests! ☺️ Code LGTM

File: src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java [hunk 9]
```diff
@@ -152,14 +171,14 @@ public void init() {
 
         ITopic<String> pauseBuildAgentTopic = hazelcastInstance.getTopic("pauseBuildAgentTopic");
         pauseBuildAgentTopic.addMessageListener(message -> {
-            if (message.getMessageObject().equals(hazelcastInstance.getCluster().getLocalMember().getAddress().toString())) {
+            if (Objects.equals(message.getMessageObject(), buildAgentShortName)) {
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java:174]: `buildAgentShortName` will never be null
```suggestion
            if (buildAgentShortName.equals(message.getMessageObject())) {
```

File: src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java [hunk 10]
```diff
@@ -152,14 +171,14 @@ public void init() {
...
                 pauseBuildAgent();
             }
         });
 
         ITopic<String> resumeBuildAgentTopic = hazelcastInstance.getTopic("resumeBuildAgentTopic");
         resumeBuildAgentTopic.addMessageListener(message -> {
-            if (message.getMessageObject().equals(hazelcastInstance.getCluster().getLocalMember().getAddress().toString())) {
+            if (Objects.equals(message.getMessageObject(), buildAgentShortName)) {
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java:181]:
```suggestion
            if (buildAgentShortName.equals(message.getMessageObject())) {
```

File: src/main/java/de/tum/cit/aet/artemis/core/web/admin/AdminBuildJobQueueResource.java [hunk 11]
```diff
@@ -95,8 +95,11 @@ public ResponseEntity<List<BuildAgentInformation>> getBuildAgentSummary() {
...
     public ResponseEntity<BuildAgentInformation> getBuildAgentDetails(@RequestParam String agentName) {
         log.debug("REST request to get information on build agent {}", agentName);
-        BuildAgentInformation buildAgentDetails = localCIBuildJobQueueService.getBuildAgentInformation().stream().filter(agent -> agent.name().equals(agentName)).findFirst()
-                .orElse(null);
+        BuildAgentInformation buildAgentDetails = localCIBuildJobQueueService.getBuildAgentInformation().stream().filter(agent -> agent.buildAgent().name().equals(agentName))
+                .findFirst().orElse(null);
+        if (buildAgentDetails == null) {
+            return ResponseEntity.notFound().build();
+        }
         return ResponseEntity.ok(buildAgentDetails);
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/core/web/admin/AdminBuildJobQueueResource.java:103]: Why not use the optional directly?
```suggestion
        Optional<BuildAgentInformation> buildAgentDetails = localCIBuildJobQueueService.getBuildAgentInformation().stream().filter(agent -> agent.buildAgent().name().equals(agentName))
                .findFirst();
        if (buildAgentDetails.isEmpty()) {
            return ResponseEntity.notFound().build();
        }
        return ResponseEntity.ok(buildAgentDetails.get());
```

File: src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java [hunk 12]
```diff
@@ -114,9 +116,15 @@ public class SharedQueueProcessingService {
      */
     private final AtomicBoolean processResults = new AtomicBoolean(true);
 
-    @Value("${artemis.continuous-integration.pause-grace-period-seconds:15}")
+    @Value("${artemis.continuous-integration.pause-grace-period-seconds:60}")
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java:119]: Please add a PR description and explain this change

File: src/main/java/de/tum/cit/aet/artemis/buildagent/dto/BuildAgent.java [hunk 13]
```diff
@@ -0,0 +1,10 @@
+package de.tum.cit.aet.artemis.buildagent.dto;
+
+import java.io.Serial;
+import java.io.Serializable;
+
+public record BuildAgent(String name, String memberAddress, String displayName) implements Serializable {
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/buildagent/dto/BuildAgent.java:6]: Why do all these DTOs not contain `DTO` in their name? I confused them for actual domain objects for a few seconds
```suggestion
public record BuildAgentDTO(String name, String memberAddress, String displayName) implements Serializable {
```

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: docs/admin/setup/distributed.rst (hunk 2 above)
Comment (florian-glombik) [reply docs/admin/setup/distributed.rst:621]: So I am fine if we leave it as it is, just wanted to make sure it is intended :)

**Review by krusche — APPROVED** [review]

**Pull request merged**